In [1]:
import pandas as pd

df = pd.read_csv(r"C:/Users/Admin/Downloads/cars_data_clean.csv.zip")

print("Shape:", df.shape)
df.info()
df.describe()
df.head()

Shape: (37813, 66)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37813 entries, 0 to 37812
Data columns (total 66 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   usedCarSkuId              37813 non-null  object 
 1   loc                       31963 non-null  object 
 2   myear                     37813 non-null  int64  
 3   body                      37794 non-null  object 
 4   transmission              37813 non-null  object 
 5   fuel                      37813 non-null  object 
 6   km                        37813 non-null  float64
 7   ip                        37813 non-null  int64  
 8   images                    37813 non-null  object 
 9   imgCount                  37813 non-null  int64  
 10  threesixty                37813 non-null  bool   
 11  dvn                       37813 non-null  object 
 12  oem                       37813 non-null  object 
 13  model                     37813 non-null  

,usedCarSkuId,loc,myear,body,transmission,fuel,km,ip,images,imgCount,...,Fuel Suppy System,Compression Ratio,Alloy Wheel Size,Ground Clearance Unladen,Max Power Delivered,Max Power At,Max Torque Delivered,Max Torque At,Bore,Stroke
0,7111bf25-97af-47f9-867b-40879190d800,gomti nagar,2016,hatchback,manual,cng,69162.0,0,[{'img': 'https://images10.gaadi.com/usedcar_i...,15,...,NaN,NaN,NaN,NaN,58.16,6200.0,77.0,3500.0,69.0,NaN
1,c309efc1-efaf-4f82-81ad-dcb38eb36665,borivali west,2015,hatchback,manual,cng,45864.0,0,[{'img': 'https://images10.gaadi.com/usedcar_i...,15,...,Gasoline Port Injection,11.0,NaN,NaN,58.20,6000.0,78.0,3500.0,73.0,NaN
2,7609f710-0c97-4f00-9a47-9b9284b62d3a,jasola,2015,sedan,manual,cng,81506.0,0,[{'img': 'https://images10.gaadi.com/usedcar_i...,15,...,NaN,NaN,NaN,NaN,86.70,6000.0,109.0,4500.0,NaN,NaN
3,278b76e3-5539-4a5e-ae3e-353a2e3b6d7d,jasola,2013,hatchback,manual,cng,115893.0,0,[{'img': ''}],0,...,Multi-Point Fuel Injection,NaN,13.0,NaN,58.20,6200.0,77.0,3500.0,NaN,NaN
4,b1eab99b-a606-48dd-a75b-57feb8a9ad92,mumbai g.p.o.,2022,muv,manual,cng,18900.0,0,[{'img': 'https://images10.gaadi.com/usedcar_i...,6,...,NaN,12.0,NaN,NaN,86.63,5500.0,121.5,4200.0,NaN,NaN


Task 1 — Data Exploration

Observed Data Quality Issues:
Missing Values
Many columns have missing values
Example: Stroke, Bore, Compression Ratio have a lot of nulls
Too Many Irrelevant Columns
Columns like images, top_features, interior_features are not useful for prediction
Inconsistent Column Names
Some columns have spaces like "Engine Type", "No of Cylinder"
Duplicate Data Possibility
Need to check for duplicates
Target Column Issue
Dataset uses listed_price instead of selling_price

Task 2 — Data Cleaning

In [6]:
df = df.rename(columns={'listed_price': 'selling_price'})
df

,loc,myear,body,transmission,fuel,km,ip,imgCount,threesixty,dvn,...,Fuel Suppy System,Compression Ratio,Alloy Wheel Size,Ground Clearance Unladen,Max Power Delivered,Max Power At,Max Torque Delivered,Max Torque At,Bore,Stroke
0,gomti nagar,2016,hatchback,manual,cng,69162.0,0,15,False,maruti wagon r lxi cng,...,NaN,NaN,NaN,NaN,58.16,6200.0,77.0,3500.0,69.0,NaN
1,borivali west,2015,hatchback,manual,cng,45864.0,0,15,False,maruti celerio green vxi,...,Gasoline Port Injection,11.0,NaN,NaN,58.20,6000.0,78.0,3500.0,73.0,NaN
2,jasola,2015,sedan,manual,cng,81506.0,0,15,False,honda amaze s plus i-vtec,...,NaN,NaN,NaN,NaN,86.70,6000.0,109.0,4500.0,NaN,NaN
3,jasola,2013,hatchback,manual,cng,115893.0,0,0,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,NaN,13.0,NaN,58.20,6200.0,77.0,3500.0,NaN,NaN
4,mumbai g.p.o.,2022,muv,manual,cng,18900.0,0,6,False,maruti ertiga vxi cng,...,NaN,12.0,NaN,NaN,86.63,5500.0,121.5,4200.0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37808,thathawade,2018,suv,manual,diesel,49882.0,0,15,False,tata nexon 1.5 revotorq xz plus,...,Common Rail Injection,NaN,16.0,209.0,108.50,3750.0,260.0,2125.0,76.0,82.5
37809,kharadi,2011,sedan,manual,diesel,82843.0,0,27,False,tata manza aura quadrajet bs iv,...,Common Rail Injection,NaN,15.0,NaN,88.80,4000.0,200.0,2375.0,NaN,NaN
37810,kothrud,2020,hatchback,manual,diesel,58000.0,0,11,False,tata altroz xz diesel,...,NaN,NaN,16.0,NaN,88.77,4000.0,200.0,2125.0,NaN,NaN
37811,kothrud,2014,hatchback,manual,diesel,62000.0,0,8,False,tata indica ev2 ls,...,Common Rail Injection,NaN,NaN,NaN,69.00,4000.0,140.0,2400.0,NaN,NaN


In [7]:
df = df.dropna(subset=['selling_price'])
df

,loc,myear,body,transmission,fuel,km,ip,imgCount,threesixty,dvn,...,Fuel Suppy System,Compression Ratio,Alloy Wheel Size,Ground Clearance Unladen,Max Power Delivered,Max Power At,Max Torque Delivered,Max Torque At,Bore,Stroke
0,gomti nagar,2016,hatchback,manual,cng,69162.0,0,15,False,maruti wagon r lxi cng,...,NaN,NaN,NaN,NaN,58.16,6200.0,77.0,3500.0,69.0,NaN
1,borivali west,2015,hatchback,manual,cng,45864.0,0,15,False,maruti celerio green vxi,...,Gasoline Port Injection,11.0,NaN,NaN,58.20,6000.0,78.0,3500.0,73.0,NaN
2,jasola,2015,sedan,manual,cng,81506.0,0,15,False,honda amaze s plus i-vtec,...,NaN,NaN,NaN,NaN,86.70,6000.0,109.0,4500.0,NaN,NaN
3,jasola,2013,hatchback,manual,cng,115893.0,0,0,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,NaN,13.0,NaN,58.20,6200.0,77.0,3500.0,NaN,NaN
4,mumbai g.p.o.,2022,muv,manual,cng,18900.0,0,6,False,maruti ertiga vxi cng,...,NaN,12.0,NaN,NaN,86.63,5500.0,121.5,4200.0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37808,thathawade,2018,suv,manual,diesel,49882.0,0,15,False,tata nexon 1.5 revotorq xz plus,...,Common Rail Injection,NaN,16.0,209.0,108.50,3750.0,260.0,2125.0,76.0,82.5
37809,kharadi,2011,sedan,manual,diesel,82843.0,0,27,False,tata manza aura quadrajet bs iv,...,Common Rail Injection,NaN,15.0,NaN,88.80,4000.0,200.0,2375.0,NaN,NaN
37810,kothrud,2020,hatchback,manual,diesel,58000.0,0,11,False,tata altroz xz diesel,...,NaN,NaN,16.0,NaN,88.77,4000.0,200.0,2125.0,NaN,NaN
37811,kothrud,2014,hatchback,manual,diesel,62000.0,0,8,False,tata indica ev2 ls,...,Common Rail Injection,NaN,NaN,NaN,69.00,4000.0,140.0,2400.0,NaN,NaN


In [9]:
drop_cols = [
    'usedCarSkuId', 'images', 'top_features', 'comfort_features',
    'interior_features', 'exterior_features', 'safety_features'
]

df = df.drop(columns=drop_cols, errors='ignore')
df

,loc,myear,body,transmission,fuel,km,ip,imgCount,threesixty,dvn,...,Fuel Suppy System,Compression Ratio,Alloy Wheel Size,Ground Clearance Unladen,Max Power Delivered,Max Power At,Max Torque Delivered,Max Torque At,Bore,Stroke
0,gomti nagar,2016,hatchback,manual,cng,69162.0,0,15,False,maruti wagon r lxi cng,...,NaN,NaN,NaN,NaN,58.16,6200.0,77.0,3500.0,69.0,NaN
1,borivali west,2015,hatchback,manual,cng,45864.0,0,15,False,maruti celerio green vxi,...,Gasoline Port Injection,11.0,NaN,NaN,58.20,6000.0,78.0,3500.0,73.0,NaN
2,jasola,2015,sedan,manual,cng,81506.0,0,15,False,honda amaze s plus i-vtec,...,NaN,NaN,NaN,NaN,86.70,6000.0,109.0,4500.0,NaN,NaN
3,jasola,2013,hatchback,manual,cng,115893.0,0,0,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,NaN,13.0,NaN,58.20,6200.0,77.0,3500.0,NaN,NaN
4,mumbai g.p.o.,2022,muv,manual,cng,18900.0,0,6,False,maruti ertiga vxi cng,...,NaN,12.0,NaN,NaN,86.63,5500.0,121.5,4200.0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37808,thathawade,2018,suv,manual,diesel,49882.0,0,15,False,tata nexon 1.5 revotorq xz plus,...,Common Rail Injection,NaN,16.0,209.0,108.50,3750.0,260.0,2125.0,76.0,82.5
37809,kharadi,2011,sedan,manual,diesel,82843.0,0,27,False,tata manza aura quadrajet bs iv,...,Common Rail Injection,NaN,15.0,NaN,88.80,4000.0,200.0,2375.0,NaN,NaN
37810,kothrud,2020,hatchback,manual,diesel,58000.0,0,11,False,tata altroz xz diesel,...,NaN,NaN,16.0,NaN,88.77,4000.0,200.0,2125.0,NaN,NaN
37811,kothrud,2014,hatchback,manual,diesel,62000.0,0,8,False,tata indica ev2 ls,...,Common Rail Injection,NaN,NaN,NaN,69.00,4000.0,140.0,2400.0,NaN,NaN


In [11]:
num_cols = df.select_dtypes(include=['int64', 'float64']).columns
df[num_cols] = df[num_cols].fillna(df[num_cols].median())

cat_cols = df.select_dtypes(include=['object']).columns
df[cat_cols] = df[cat_cols].fillna(df[cat_cols].mode().iloc[0])
df

,loc,myear,body,transmission,fuel,km,ip,imgCount,threesixty,dvn,...,Fuel Suppy System,Compression Ratio,Alloy Wheel Size,Ground Clearance Unladen,Max Power Delivered,Max Power At,Max Torque Delivered,Max Torque At,Bore,Stroke
0,gomti nagar,2016,hatchback,manual,cng,69162.0,0,15,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,10.5,16.0,190.0,58.16,6200.0,77.0,3500.0,69.0,85.0
1,borivali west,2015,hatchback,manual,cng,45864.0,0,15,False,maruti celerio green vxi,...,Gasoline Port Injection,11.0,16.0,190.0,58.20,6000.0,78.0,3500.0,73.0,85.0
2,jasola,2015,sedan,manual,cng,81506.0,0,15,False,honda amaze s plus i-vtec,...,Multi-Point Fuel Injection,10.5,16.0,190.0,86.70,6000.0,109.0,4500.0,73.0,85.0
3,jasola,2013,hatchback,manual,cng,115893.0,0,0,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,10.5,13.0,190.0,58.20,6200.0,77.0,3500.0,73.0,85.0
4,mumbai g.p.o.,2022,muv,manual,cng,18900.0,0,6,False,maruti ertiga vxi cng,...,Multi-Point Fuel Injection,12.0,16.0,190.0,86.63,5500.0,121.5,4200.0,73.0,85.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37808,thathawade,2018,suv,manual,diesel,49882.0,0,15,False,tata nexon 1.5 revotorq xz plus,...,Common Rail Injection,10.5,16.0,209.0,108.50,3750.0,260.0,2125.0,76.0,82.5
37809,kharadi,2011,sedan,manual,diesel,82843.0,0,27,False,tata manza aura quadrajet bs iv,...,Common Rail Injection,10.5,15.0,190.0,88.80,4000.0,200.0,2375.0,73.0,85.0
37810,kothrud,2020,hatchback,manual,diesel,58000.0,0,11,False,tata altroz xz diesel,...,Multi-Point Fuel Injection,10.5,16.0,190.0,88.77,4000.0,200.0,2125.0,73.0,85.0
37811,kothrud,2014,hatchback,manual,diesel,62000.0,0,8,False,tata indica ev2 ls,...,Common Rail Injection,10.5,16.0,190.0,69.00,4000.0,140.0,2400.0,73.0,85.0


In [12]:
df['oem'] = df['oem'].str.strip().str.lower()
df

,loc,myear,body,transmission,fuel,km,ip,imgCount,threesixty,dvn,...,Fuel Suppy System,Compression Ratio,Alloy Wheel Size,Ground Clearance Unladen,Max Power Delivered,Max Power At,Max Torque Delivered,Max Torque At,Bore,Stroke
0,gomti nagar,2016,hatchback,manual,cng,69162.0,0,15,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,10.5,16.0,190.0,58.16,6200.0,77.0,3500.0,69.0,85.0
1,borivali west,2015,hatchback,manual,cng,45864.0,0,15,False,maruti celerio green vxi,...,Gasoline Port Injection,11.0,16.0,190.0,58.20,6000.0,78.0,3500.0,73.0,85.0
2,jasola,2015,sedan,manual,cng,81506.0,0,15,False,honda amaze s plus i-vtec,...,Multi-Point Fuel Injection,10.5,16.0,190.0,86.70,6000.0,109.0,4500.0,73.0,85.0
3,jasola,2013,hatchback,manual,cng,115893.0,0,0,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,10.5,13.0,190.0,58.20,6200.0,77.0,3500.0,73.0,85.0
4,mumbai g.p.o.,2022,muv,manual,cng,18900.0,0,6,False,maruti ertiga vxi cng,...,Multi-Point Fuel Injection,12.0,16.0,190.0,86.63,5500.0,121.5,4200.0,73.0,85.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37808,thathawade,2018,suv,manual,diesel,49882.0,0,15,False,tata nexon 1.5 revotorq xz plus,...,Common Rail Injection,10.5,16.0,209.0,108.50,3750.0,260.0,2125.0,76.0,82.5
37809,kharadi,2011,sedan,manual,diesel,82843.0,0,27,False,tata manza aura quadrajet bs iv,...,Common Rail Injection,10.5,15.0,190.0,88.80,4000.0,200.0,2375.0,73.0,85.0
37810,kothrud,2020,hatchback,manual,diesel,58000.0,0,11,False,tata altroz xz diesel,...,Multi-Point Fuel Injection,10.5,16.0,190.0,88.77,4000.0,200.0,2125.0,73.0,85.0
37811,kothrud,2014,hatchback,manual,diesel,62000.0,0,8,False,tata indica ev2 ls,...,Common Rail Injection,10.5,16.0,190.0,69.00,4000.0,140.0,2400.0,73.0,85.0


In [13]:
print("Duplicates:", df.duplicated().sum())
df = df.drop_duplicates()

Duplicates: 19


In [15]:
df = df[df['selling_price'] > 0]
df

,loc,myear,body,transmission,fuel,km,ip,imgCount,threesixty,dvn,...,Fuel Suppy System,Compression Ratio,Alloy Wheel Size,Ground Clearance Unladen,Max Power Delivered,Max Power At,Max Torque Delivered,Max Torque At,Bore,Stroke
0,gomti nagar,2016,hatchback,manual,cng,69162.0,0,15,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,10.5,16.0,190.0,58.16,6200.0,77.0,3500.0,69.0,85.0
1,borivali west,2015,hatchback,manual,cng,45864.0,0,15,False,maruti celerio green vxi,...,Gasoline Port Injection,11.0,16.0,190.0,58.20,6000.0,78.0,3500.0,73.0,85.0
2,jasola,2015,sedan,manual,cng,81506.0,0,15,False,honda amaze s plus i-vtec,...,Multi-Point Fuel Injection,10.5,16.0,190.0,86.70,6000.0,109.0,4500.0,73.0,85.0
3,jasola,2013,hatchback,manual,cng,115893.0,0,0,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,10.5,13.0,190.0,58.20,6200.0,77.0,3500.0,73.0,85.0
4,mumbai g.p.o.,2022,muv,manual,cng,18900.0,0,6,False,maruti ertiga vxi cng,...,Multi-Point Fuel Injection,12.0,16.0,190.0,86.63,5500.0,121.5,4200.0,73.0,85.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37808,thathawade,2018,suv,manual,diesel,49882.0,0,15,False,tata nexon 1.5 revotorq xz plus,...,Common Rail Injection,10.5,16.0,209.0,108.50,3750.0,260.0,2125.0,76.0,82.5
37809,kharadi,2011,sedan,manual,diesel,82843.0,0,27,False,tata manza aura quadrajet bs iv,...,Common Rail Injection,10.5,15.0,190.0,88.80,4000.0,200.0,2375.0,73.0,85.0
37810,kothrud,2020,hatchback,manual,diesel,58000.0,0,11,False,tata altroz xz diesel,...,Multi-Point Fuel Injection,10.5,16.0,190.0,88.77,4000.0,200.0,2125.0,73.0,85.0
37811,kothrud,2014,hatchback,manual,diesel,62000.0,0,8,False,tata indica ev2 ls,...,Common Rail Injection,10.5,16.0,190.0,69.00,4000.0,140.0,2400.0,73.0,85.0


In [16]:
print(df.shape)
df.head()

(37794, 59)


,loc,myear,body,transmission,fuel,km,ip,imgCount,threesixty,dvn,...,Fuel Suppy System,Compression Ratio,Alloy Wheel Size,Ground Clearance Unladen,Max Power Delivered,Max Power At,Max Torque Delivered,Max Torque At,Bore,Stroke
0,gomti nagar,2016,hatchback,manual,cng,69162.0,0,15,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,10.5,16.0,190.0,58.16,6200.0,77.0,3500.0,69.0,85.0
1,borivali west,2015,hatchback,manual,cng,45864.0,0,15,False,maruti celerio green vxi,...,Gasoline Port Injection,11.0,16.0,190.0,58.20,6000.0,78.0,3500.0,73.0,85.0
2,jasola,2015,sedan,manual,cng,81506.0,0,15,False,honda amaze s plus i-vtec,...,Multi-Point Fuel Injection,10.5,16.0,190.0,86.70,6000.0,109.0,4500.0,73.0,85.0
3,jasola,2013,hatchback,manual,cng,115893.0,0,0,False,maruti wagon r lxi cng,...,Multi-Point Fuel Injection,10.5,13.0,190.0,58.20,6200.0,77.0,3500.0,73.0,85.0
4,mumbai g.p.o.,2022,muv,manual,cng,18900.0,0,6,False,maruti ertiga vxi cng,...,Multi-Point Fuel Injection,12.0,16.0,190.0,86.63,5500.0,121.5,4200.0,73.0,85.0


Task 3 — Baseline MAE

In [17]:
mean_price = df['selling_price'].mean()
print("Mean Price:", mean_price)

Mean Price: 800095.2065142615


In [20]:
df['baseline_pred'] = mean_price

In [21]:
from sklearn.metrics import mean_absolute_error

mae = mean_absolute_error(df['selling_price'], df['baseline_pred'])

print("Baseline MAE:", mae)


Baseline MAE: 551116.1133220675


The dataset does not contain a mileage column, so the mileage step was not applicable.
The dataset was cleaned by handling missing values, removing duplicates, and standardizing columns.
A baseline model was created using the mean selling price.
The Mean Absolute Error (MAE) obtained is 551116.1133220675.
This baseline serves as a reference for future machine learning models.